pkg/chains/formats/slsa/v2alpha2/internal/resolved_dependencies/resolved_dependencies.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2023 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resolveddependencies
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/common"
+	v1 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v1"
+	"github.com/tektoncd/chains/pkg/chains/formats/slsa/internal/material"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	// pipelineConfigName is the name of the resolved dependency of the pipelineRef.
+	pipelineConfigName = "pipeline"
+	// taskConfigName is the name of the resolved dependency of the top level taskRef.
+	taskConfigName = "task"
+	// pipelineTaskConfigName is the name of the resolved dependency of the pipeline task.
+	pipelineTaskConfigName = "pipelineTask"
+	// inputResultName is the name of the resolved dependency generated from Type hinted parameters or results.
+	inputResultName = "inputs/result"
+	// pipelineResourceName is the name of the resolved dependency of pipeline resource.
+	pipelineResourceName = "pipelineResource"
+)
+
+// TaskRun constructs `predicate.resolvedDependencies` section by collecting all the artifacts that influence a taskrun such as source code repo and step&sidecar base images.
+func TaskRun(ctx context.Context, tro *objects.TaskRunObject) ([]v1.ResourceDescriptor, error) {
+	var resolvedDependencies []v1.ResourceDescriptor
+	var err error
+
+	// add top level task config
+	if p := tro.Status.Provenance; p != nil && p.RefSource != nil {
+		rd := v1.ResourceDescriptor{
+			Name:   taskConfigName,
+			URI:    p.RefSource.URI,
+			Digest: p.RefSource.Digest,
+		}
+		resolvedDependencies = append(resolvedDependencies, rd)
+	}
+
+	mats := []common.ProvenanceMaterial{}
+
+	// add step and sidecar images
+	if err := material.AddStepImagesToMaterials(tro.Status.Steps, &mats); err != nil {
```

Comment:
I'm a bit concerned that some of the functions in `material` take a pointer address to a materials slice, like this one, for two reasons:

1. It seems bad practice to pass a slice pointer around.
2. Other functions on the `materials` package deal with adding things to materials but don't take a pre-existing slice, e.g. `AddMaterialsFromTaskParamsAndResults`.

How about something like this?

```
allMaterials := []common.ProvenanceMaterial{}

mats, err := material.FromStepImages(tro.Status.Steps)
if err != nil {...}
allMaterials = append(allMaterials, mats...)
```

Reply:
Agreed! I think we want to consolidate this. @chuangw6 pointed this out as well. But its better left as its own cleanup PR which I will do after this.